core/src/main/scala/Diet.scala

```diff
@@ -0,0 +1,321 @@
+          if (discrete.adj(y, value))
+            joinRight(DietNode(x, value, l, r))
+          else
+            DietNode(x, y, l, r.add(value))
+      }
+      else d
+    }
+  }
+
+  /**
+   * remove x from the tree
+   */
+  def remove(x: A)(implicit discrete: Discrete[A]): Diet[A] = this match {
+    case EmptyDiet()  =>  this
+    case DietNode(a, b, l, r) =>  {
+      if (discrete.compare(x, a) == LT) {
+        DietNode(a, b, l.remove(x), r)
+      }
+      else if (discrete.compare(x, b) == GT) {
+        DietNode(a, b, l, r.remove(x))
+      }
+      else if (discrete.compare(x, a) == EQ) {
+        if (discrete.compare(a, b) == EQ)
+          merge(l, r)
+        else
+          DietNode(discrete.succ(a), b, l, r)
+      }
+      else if (discrete.compare(x, b) == EQ) {
+        DietNode(a, discrete.pred(b), l, r)
+      }
+      else {
+        DietNode(a, discrete.pred(x), l, DietNode(discrete.succ(x), b, EmptyDiet(), r))
+      }
+    }
+  }
+
+  /**
+   * alias for add
+   */
+  def +(value: A)(implicit discrete: Discrete[A]): Diet[A] = add(value)
+
+  /**
+   * alias for remove
+   */
+  def -(value: A)(implicit discrete: Discrete[A]): Diet[A] = remove(value)
+
+  /**
+   * min value in the tree
+   */
+  def min: Option[A] = this match {
+    case EmptyDiet()  =>  None()
+    case DietNode(x, _, EmptyDiet(), _) => Some(x)
+    case DietNode(_, _, l, _) => l.min
+  }
+
+  /**
+   * max value in the tree
+   */
+  def max: Option[A] = this match {
+    case EmptyDiet()  => None()
+    case DietNode(_, y, _, EmptyDiet()) => Some(y)
+    case DietNode(_, _, _, r) => r.max
+  }
+
+  def map[B](f: A => B)(implicit discrete: Discrete[B]): Diet[B] = this match {
+    case EmptyDiet()          =>  Diet.empty[B]()
+    case DietNode(a, b, l, r) =>  {
+      val (lp, rp) = (l.map(f), r.map(f))
+
+      val n = lp + f(a) + f(b)
+
+      merge(n, rp)
+    }
+  }
+
+  def foldRight[B](s: B)(f: (B, A) => B)(implicit discrete: Discrete[A]): B = this match {
+    case EmptyDiet()          =>  s
+    case DietNode(a, b, l, r) =>  l.foldRight(DRange(a, b).reverse().foldLeft(r.foldRight(s)(f))(f))(f)
+  }
+
+  def foldLeft[B](s: B)(f: (B, A) => B)(implicit discrete: Discrete[A]): B = this match {
+    case EmptyDiet()          =>  s
+    case DietNode(a, b, l, r) =>  r.foldLeft(DRange(a, b).generate().foldLeft(l.foldLeft[B](s)(f))(f))(f)
+  }
+
+}
+
+object Diet {
+  def apply[A](): Diet[A] = empty()
+
+  def empty[A](): Diet[A] = EmptyDiet()
+
+  private [dogs] case class DietNode[A](x: A, y: A, left: Diet[A], right: Diet[A]) extends Diet[A] {
+    override val isEmpty: Boolean = false
+  }
+
+  private [dogs] case object EmptyDiet extends Diet[Nothing] {
+    def unapply[A](diet: Diet[A]): Boolean = diet.isEmpty
+
+    def apply[A](): Diet[A] = this.asInstanceOf[Diet[A]]
+
+    override val isEmpty: Boolean = true
+  }
+
+  /**
+   * merge to Diets
+   */
+  private [dogs] def merge[A](l: Diet[A], r: Diet[A]): Diet[A] = (l, r) match {
```

Comment:
No, it does not merge overlapping Diet(s). It is intended for remove. It is something to look at in the future.